README.md
# typer.js
## What is typer.js
**typer.js** is a framework that allows to type automatically if you have put a value in the tag.

## Installation
You can use typer.js via CDN
- CDN:
  - Normal: https://jsdelivr.net/gh/Dob6458/typerjs/libs/typer.js
  - Minified: https://jsdelivr.net/gh/Dob6458/typerjs/libs/typer.min.js

## Changelogs
You can check it in [CHANGELOG.md](CHANGELOG.md) file in this repository

## Code
Once **typer.js** installed, import it using `<script>` tag on HTML file:
### Importation
```html
<!--
  Use this code in <head> or <body> tag
-->
<script src="typer.js"></script>
<script src="index.js"></script>

<!--
  Use <span> or <div> tags
-->
<span id="typer"></span>
```

Then, create a file called `index.js` file and copy this code.
### Simple code
```js
window.onload = function() {
  /**
    You can use the id, class or tags through "document.querySelector()"
  */
  var typer = new typer("#typer", {
    string: "typer.js is working successfuly"
  });
}
```
Automatically, this selected element is gonna type!

#### Result:

![result_one](images/image_one.gif)

### typer with options
Of course, you can use options to edit something:
```js
window.onload = function() {
  // There's options in this code
  var typer = new typer("#typer", {
    color: "orange",
    speed: 40,
    transition: 150,
    string: "typer.js is working successfuly"
  });
}
```
#### Result:

![result_two](images/image_two.gif)

We have created this options through `Object.assign({}, options_default, options)`
### Events
```js
window.onload = function() {
  var typer = new typer("#typer");

  typer.on("click", function() {
    // Code here
  });
}
```
We use `document.addEventListener(event, callback)` for the events if `typer()` class is declared.

For more information, you can go to the [documentation](https://dob6458.github.io/libraries/typerjs/docs) (Website not published yet)

## Contributors
Thanks to:

![Seniru](https://avatars0.githubusercontent.com/u/34127015?s=50&v=4)

## Issues or Feedback
If you have a problem, please, create an [issue](https://github.com/Dob6458/typerjs/issues) on Github or contact me on:
- Discord: // ζohιπ#5914
- Sololearn Profile ID: 6015460
- Email: dob6458@outlook.com

CHANGELOG.md
# Changelogs

Coming soon for the future versions...

## 1.0.4
- Renamed **Typer.js** instead **typing.js**
- Updated README.md

## 1.0.3
- Added promise to callback private functions
- Added `string` variable in the options
- Code fixed

## 1.0.2
- Adding docs and demos and publishing it in Github Pages (failed)

## 1.0.1
- Code fixed
- Functions added `on()` and `off()`
- Private functions worked correctly
- CDN avaliable (check it on README.md)

## 1.0 (Released)
- Project created
- You can use options in this class as `new Typing("#typing", {<options>})`

tests/index.js
window.onload = function() {
  // There's options in this code
  var options = {
    color: "black",
    speed: 60,
    transition: 150, 
    width: 1.5,
    string: "Typing.js is working successfuly"
  }

  var typing = new Typing("#typing", options);
}

libs/typer.js
//
// Library created by // Zohir on 02/09/2019
// Typer.js (Dob6458/typerjs)
//

function Typer(id, options={}) {
  ///////////////////// CONSTRUCTOR ZONE /////////////////////
  var element = document.querySelector(id);
  
  // Make the options by default
  options_default = {
    color: "black",
    speed: 60,
    transition: 150, 
    width: 1.5,
    string: ""
  };

  // Assign it to make options
  let actual = Object.assign({}, options_default, options);
  ///////////////////// CONSTRUCTOR ZONE /////////////////////

  // Create a promise
  var promise = new Promise((resolve, reject) => {
    resolve("Function work");
    reject("Function not work");
  });

  // Callback function if this promise is work
  promise.then(() => {
    init(element); typing(element); booleans(element);
  });
  
  // When the promise is not work
  promise.catch((err) => {
    console.log(err);
  })

  /**
   * This function allows to add event when typing
   * @public This function is public
   * @param {string} event "event" must be a string
   * @param {function} callback "callback" must be a function
   */
  this.on = function(event, callback) {
    // If "event" variable is not a string
    if (typeof event !== "string") {
      return console.error("The event must be in string");
    }

    // If "callback" variable is not a function
    if (typeof callback !== "function") {
      return console.error("The callback must be a function");
    }

    // Add event if it's work
    return element.addEventListener(event, callback);
  };

  /**
   * This allows to remove event when typing
   * @public This function is public
   * @param {string} event "event" must be a string
   * @param {function} callback "callback" must be a function
   * @param {boolean} bool "bool" is null (by default)
   */
  this.off = function(event, callback, bool=null) {
    // If "event" variable is not a string
    if (typeof event !== "string") {
      return console.error("The event must be a string");
    };

    // If "callback" variable is not a function
    if (typeof callback !== "function") {
      return console.error("The callback must be a function");
    };

    // Remove event if it's work
    return element.removeEventListener(event, callback, bool);
  };

  /**
   * Create a function that allows to start type automatically
   * @private This function is private
   * @param {object} element 
   */
  let init = function(element) {
    let value = actual.string;
    let type = "";

    let position = 0;

    let x = setInterval(function() {
      type += value[position];
      element.innerText = type;

      position++;

      if (position > value.length - 1) {
        clearInterval(x);
      }
    }, actual.speed);
  };
  
  /**
   * @private This function is private
   * @param {object} element 
   */
  let typing = function(element) {
    let borders = actual.width + "solid " + actual.color;
    element.style.transition = actual.transition;
    element.style.borderRight = borders;
    element.style.paddingRight = "1px";
    element.style.display = "inline";
  };
  
  /**
   * Create a function that allows to animate the cursor while typing
   * 
   * @private This function is private
   * @param {object} element
   */
  let booleans = function(element) {
    let bools = false;
          
    setInterval(function() {
      if (bools) {
        bools = false;
        element.style.borderRight = actual.width + "px solid transparent";
      } else {
        bools = true;
        element.style.borderRight = actual.width + "px solid " + actual.color;
      }
    }, actual.transition * 3);
  };
}


libs/typing.js
//
// Library created by // Zohir on 02/09/2019
// Typing.js (Dob6458/typingjs)
//

function Typing(id, options={}) {
  ///////////////////// CONSTRUCTOR ZONE /////////////////////
  var element = document.querySelector(id);
  
  // Make the options by default
  options_default = {
    color: "black",
    speed: 60,
    transition: 150, 
    width: 1.5,
    string: ""
  };

  // Assign it to make options
  let actual = Object.assign({}, options_default, options);
  ///////////////////// CONSTRUCTOR ZONE /////////////////////

  // Create a promise
  var promise = new Promise((resolve, reject) => {
    resolve("Function work");
    reject("Function not work");
  });

  // Callback function if this promise is work
  promise.then(() => {
    init(element); typing(element); booleans(element);
  });
  
  // When the promise is not work
  promise.catch((err) => {
    console.log(err);
  })

  /**
   * This function allows to add event when typing
   * @public This function is public
   * @param {string} event "event" must be a string
   * @param {function} callback "callback" must be a function
   */
  this.on = function(event, callback) {
    // If "event" variable is not a string
    if (typeof event !== "string") {
      return console.error("The event must be in string");
    }

    // If "callback" variable is not a function
    if (typeof callback !== "function") {
      return console.error("The callback must be a function");
    }

    // Add event if it's work
    return element.addEventListener(event, callback);
  };

  /**
   * This allows to remove event when typing
   * @public This function is public
   * @param {string} event "event" must be a string
   * @param {function} callback "callback" must be a function
   * @param {boolean} bool "bool" is null (by default)
   */
  this.off = function(event, callback, bool=null) {
    // If "event" variable is not a string
    if (typeof event !== "string") {
      return console.error("The event must be a string");
    };

    // If "callback" variable is not a function
    if (typeof callback !== "function") {
      return console.error("The callback must be a function");
    };

    // Remove event if it's work
    return element.removeEventListener(event, callback, bool);
  };

  /**
   * Create a function that allows to start type automatically
   * @private This function is private
   * @param {object} element 
   */
  let init = function(element) {
    let value = actual.string;
    let type = "";

    let position = 0;

    let x = setInterval(function() {
      type += value[position];
      element.innerText = type;

      position++;

      if (position > value.length - 1) {
        clearInterval(x);
      }
    }, actual.speed);
  };
  
  /**
   * @private This function is now private
   * @param {object} element 
   */
  let typing = function(element) {
    let borders = actual.width + "solid " + actual.color;
    element.style.transition = actual.transition;
    element.style.borderRight = borders;
    element.style.paddingRight = "1px";
    element.style.display = "inline";
  };
  
  /**
   * Create a function that allows to animate the cursor while typing
   * 
   * @private This function is private
   * @param {object} element
   */
  let booleans = function(element) {
    let bools = false;
          
    setInterval(function() {
      if (bools) {
        bools = false;
        element.style.borderRight = actual.width + "px solid transparent";
      } else {
        bools = true;
        element.style.borderRight = actual.width + "px solid " + actual.color;
      }
    }, actual.transition * 3);
  };
}
